LFS-260: As an administrator, I can configure the quick search widget

veronikaslc opened the pull request:
This PR is based on the [LFS-252](https://github.com/ccmbioinfo/lfs/pull/263)

Commit:
LFS-260: As an administrator, I can configure the quick search widget
Work in progress; TODO- add Subjects and Questionnaires to the backend quick search

Commit:
LFS-260: As an administrator, I can configure the quick search widget
Added Subjects and Questionnaires to the backend quick search

Commit:
LFS-260: As an administrator, I can configure the quick search widget
Added scroll to Questionnaires question when clicked on the quick search

Commit:
LFS-260: As an administrator, I can configure the quick search widget
Fixed small error in Questinnaires search

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Updated the UIExtension declaration to match the new dev state after LFS-546

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -100,15 +103,21 @@ public void init(Bindings bindings)
-                results = quickSearch(this.urlDecode(quickQuery));
+                // get the admin quick search defined params if any
+                Resource settings = getSearchSettings("quick");
+                limit = settings.getValueMap().get("limit", limit);
```
This should be the default value used in `getLongValueOrDefault`, but the way it is now it is always enforced without any way to request a different number of results. All quick queries now return 5 results, not just the quick search bar.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -412,7 +577,9 @@ private void addSummary(final JsonObjectBuilder jsonGen, final SlingHttpServletR
         jsonGen.add("limit", limits[1]);
         jsonGen.add("returnedrows", limits[2]);
-        jsonGen.add("totalrows", limits[3]);
+        if (showTotalRows) {
```
This should be a frontend setting, not a backend one. Disabling "Show the total number of results" has unintended consequences, like breaking the search results livetable which now shows `1-NaN of undefined` as the currently displayed items.

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/content/QuickSearchResultsWidget.json:
```diff
@@ -0,0 +1,10 @@
+{
```
This should be in `/apps/lfs/config/QuickSearch`.

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/QuickSearchResultsWidget.jsx:
```diff
@@ -0,0 +1,190 @@
+
+  // Fetch saved admin settings
+  let getQuickSearchSettings = () => {
+    fetch('/query?query=' + encodeURIComponent(`select * from [lfs:QuickSearchResultsWidget] as n WHERE n.'label'='quick'`))
```
Why do you query for it instead of getting it directly? You should know (and enforce) the path to it. Worse, anyone could create a new node of this type and set bogus values, which would break the widget for everyone else.

sdumitriu replied:
```suggestion
    fetch('/apps/lfs/config/QuickSearch')
```

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/QueryBuilder.java:
```diff
@@ -297,19 +268,80 @@ private JsonObject addMatchMetadata(String resourceValue, String query, String q
      * @return the content matching the query
      */
-    private Iterator<JsonObject> quickSearch(String query) throws RepositoryException
+    private Iterator<JsonObject> quickSearch(String query, long limit, String[] resourceTypes,
```
The `quick` method doesn't just apply to the quick search widget; it is named this way because it allows running a "quick" free text query, as opposed to a full JCR or Lucene query. So, most of the configuration should be handled in the frontend, not in the backend.

Commit:
LFS-260: As an administrator, I can configure the quick search widget
Addressed codereview comments

sashaandjic commented:
Error when trying to do the search
![image](https://user-images.githubusercontent.com/23243244/94188755-d5d08f80-fe77-11ea-952d-a34177f5b14a.png)

sdumitriu committed:
[misc] Added more error logging

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Gracefully handle the case when no targeted resource types is sent

Commit:
LFS-260: As an administrator, I can configure the quick search widget
Fixed params parsing on quick search query

sashaandjic commented:
Looks and works good. 
- SUBMIT button should be re-labelled to SAVE button
- I am not sure about the Questionnaires option ( How do I test it? when I searched for a specific questionnaire field I got no results)
![image](https://user-images.githubusercontent.com/23243244/94288737-c3149400-ff25-11ea-9326-6777e638a7a3.png)

veronikaslc committed:
LFS-260: As an administrator, I can configure the quick search widget

Renamed Submit button to Save

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/94295523-5c947380-ff2f-11ea-8589-2ba3faf222a6.png)

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Moved the sources in the homepage module

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Pass the allowed resource types as an actual array of values instead of a serialized one

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Bugfix: sometimes searching matches a section condition, which should not be returned

sdumitriu committed:
[cleanup] Replaced tabs with spaces

sdumitriu committed:
[misc] Declare used dependency in the pom

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Made the quick search per-resourcetype query implementation extensible

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
There's only one quick search configuration

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Add a SearchParameters for holding all the parameters, to reduce method argument proliferation and add flexibility

sdumitriu committed:
[misc] Minor code cleanup

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Better filtering of engines

sdumitriu committed:
LFS-260: As an administrator, I can configure the quick search widget
Always show link to the search results page
